### ISSUE TITLE
Regression in Type Inference for functions derived from typeclasses in TypeScript 5.5.x

### ISSUE THREAD

**Issue by gcanti** [Effect-TS/effect#3171]
### What version of Effect is running?

3.4.7 (and `@effect/typeclass@0.24.39`)

### What steps can reproduce the bug?

```ts
import * as OptionInstances from "@effect/typeclass/data/Option";
import * as RecordInstances from "@effect/typeclass/data/Record";

const traverse = RecordInstances.Traversable.traverse(OptionInstances.Applicative);
```

### What is the expected behavior?

In TypeScript version 5.4.x, hovering over `traverse` displays the following type information:

```ts
const traverse: {
    <A, R, O, E, B>(f: (a: A) => Option<B>): <TR, TO, TE>(self: ReadonlyRecord<string, A>) => Option<ReadonlyRecord<string, B>>;
    <TR, TO, TE, A, R, O, E, B>(self: ReadonlyRecord<string, A>, f: (a: A) => Option<B>): Option<ReadonlyRecord<string, B>>;
}
```

### What do you see instead?

After updating to TypeScript 5.5.x, the hover information shows a more complex and less intuitive type definition:

```ts
const traverse: {
    <A, R, O, E, B>(f: (a: A) => Kind<OptionTypeLambda, R, O, E, B>): <TR, TO, TE>(self: Kind<ReadonlyRecordTypeLambda<string>, TR, TO, TE, A>) => Kind<OptionTypeLambda, R, O, E, Kind<ReadonlyRecordTypeLambda<...>, TR, TO, TE, B>>;
    <TR, TO, TE, A, R, O, E, B>(self: ReadonlyRecord<...>, f: (a: A) => Kind<OptionTypeLambda, R, O, E, B>): Kind<OptionTypeLambda, R, O, E, Kind<ReadonlyRecordTypeLambda<...>, TR, TO, TE, B>>;
}
```

**Comment by mikearnaldi** [comment]
Potentially related to https://github.com/microsoft/TypeScript/issues/58937

**Comment by mikearnaldi** [comment]
Also another issue seems to be present:

```ts
import type { Kind } from "effect/HKT"
import type { OptionTypeLambda } from "effect/Option"

export type ShouldBeOption<R, O, E, B> = Kind<OptionTypeLambda, R, O, E, B>
// None<B> | Some<B>
```

previously Option was preserved

**Comment by mikearnaldi** [comment]
Also Unify on Option is broken: 

```ts
import type { Option, Unify } from "effect"

type Broken = Unify.Unify<Option.Option<number> | Option.Option<string>>
// Option.None<string | number> | Option.Some<string | number>
```

**Comment by ecyrbe** [comment]
Confirm, those libraries relying on HKT interface encoding are broken :
- Fastify Type Providers (1.6M downloads / week)
- Zodios v11 (50k downloads / week)
- Hotscript (132k downloads / week)
- Typeskell (just for learning purpose)

**Comment by jakebailey** [comment]
This is solely type display, yes? The code compiles the same? It's not a "regression in inference" if so. Not clear what "broken" means otherwise if you're saying this affects all of those libraries.

Consider using https://www.npmjs.com/package/every-ts to check to see where these changed.

**Comment by mikearnaldi** [comment]
> This is solely type display, yes? The code compiles the same? It's not a "regression in inference" if so. Not clear what "broken" means otherwise if you're saying this affects all of those libraries.
> 
> Consider using https://www.npmjs.com/package/every-ts to check to see where these changed.

the code compiles but the types produced are more complex than they were, which would make using those features no longer practical

**Comment by ecyrbe** [comment]
It's appears it also changes resolution on inference.
Since types are not infered (aka simplified), the HKT encoding that could stay internal to the library now need also to be exported.
Like this :
![image](https://github.com/Effect-TS/effect/assets/633115/ed85ca47-5483-4351-a24a-15143a8ee4bd)

An like @mikearnaldi  this makes those lib unusable.

**Comment by ecyrbe** [comment]
And since inference is not simplified, typescript can't do proper composition of many HKT together.
So it's definitely not just `display`
Exemple:
![image](https://github.com/Effect-TS/effect/assets/633115/a4f73bca-aaa9-4ca7-8c9f-9812ae4c0bfc)

**Comment by jakebailey** [comment]
If this has any type checking affect, it won't be https://github.com/microsoft/TypeScript/issues/58937. A bug report / repro / bisect would be appreciated.

**Comment by ecyrbe** [comment]
yes, i'll try to bisect this week-end to help you trouble shoot.

**Comment by mikearnaldi** [comment]
> If this has any type checking effect, it shouldn't be [microsoft/TypeScript#58937](https://github.com/microsoft/TypeScript/issues/58937). A bug report / repro / bisect would be appreciated.

That seems to be specific to ReturnType while this is more generic. Note that the same issue that @ecyrbe brought up applies there, if the function argument of ReturnType includes non exported symbols not computing it would lead to the same issue

**Comment by adrian-gierakowski** [comment]
Got here via twitter: does this negatively affect high level users of effect (who don’t use typeclass module directly)? Asking since I still haven’t upgraded tsc 5.4 => 5.5 and wondering if I should wait until this is resolved. Thanks!

**Comment by mikearnaldi** [comment]
> Got here via twitter: does this negatively affect high level users of effect (who don’t use typeclass module directly)? Asking since I still haven’t upgraded tsc 5.4 => 5.5 and wondering if I should wait until this is resolved. Thanks!

No it shouldn't, it only affect edge cases like match on options but nothing that can't be fixed with some explicit types

**Comment by jfet97** [comment]
@jakebailey I tried `every-ts` on the example issued by @gcanti and if I've done everything correctly it is pointing exactly at #58546 like the other issue.

**Comment by jakebailey** [comment]
The hover stuff I understand, but I'm more concerned about specific new type errors as previously shown by @ecyrbe.

**Comment by jfet97** [comment]
@jakebailey I tried bisecting typeskell and I got the same problem @ecyrbe was talking about and the culprit seems to be https://github.com/microsoft/TypeScript/pull/58546 again

If @ecyrbe could confirm this would be better

**Comment by jfet97** [comment]
@mikearnaldi I tried your examples too but they seem to be __unrelated__ with the issue we are talking about here. Please check it out, I don't see `Option` being preserved before `2b4e7e3`.

**Comment by jakebailey** [comment]
To be clear, I'm not referring to any "cannot be named" errors (which are yet another class of problem to be reported; this thread has a lot of different 5.5 things happening all at once), but type errors in https://github.com/Effect-TS/effect/issues/3171#issuecomment-2211347938 which imply that this change was not simply a type printing change. All of these should definitely reported with repros (which I don't see besides the original posting, which really is just a type display / type annotation reuse issue).

**Comment by mikearnaldi** [comment]
> @mikearnaldi I tried your examples too but they seem to be **unrelated** with the issue we are talking about here. Please check it out, I don't see `Option` being preserved before `2b4e7e3`.

I can confirm, I remembered it working but can't find a version where it does, I guess we can then separate the two issues. Maybe @Andarist can help investigating why the option alias gets lost

**Comment by ecyrbe** [comment]
> If @ecyrbe could confirm this, it would be better :)

Sorry, i have a hard time with every-ts, getting download errors non stop. But you analysis seams good.
We should give your analysis to Issue opened by david here :
https://github.com/microsoft/TypeScript/issues/58937

**Comment by jfet97** [comment]
@ecyrbe probably is a good idea to open another issue on TS specifically for HKTs with various repros of broken stuff like what Jake was asking for, not only type printing problems. I'm not sure how to get a type error from @gcanti's example tho, so I'll leave it to him (or to @mikearnaldi ). I think you should do the same with HotScript, whereas idk who could do this for [fastify-type-provider-typebox](https://github.com/fastify/fastify-type-provider-typebox), I don't even know if they've already encountered any problem.

I'm not a maintainer of any of the affected projects tho, so count me out of this decision. It's just a suggestion to avoid polluting the David's issue.

**Comment by Andarist** [comment]
> Maybe @Andarist can help investigating why the option alias gets lost

I could but I would need some extra info about this whole situation.

**Comment by jfet97** [comment]
@Andarist the situation:

<img width="500" alt="image" src="https://github.com/Effect-TS/effect/assets/24520167/14e39323-2cc5-4230-afb0-d703d34f368b">

**Comment by jakebailey** [comment]
I spent a little bit of time trying to figure out what was going on here and sent a potential fix (which I'm not 100% certain of). Try using the build at https://github.com/microsoft/TypeScript/pull/59170#issuecomment-2212915200 and see if that seems to fix things for your cases.

No guarantees, this PR is extremely unscientific.

**Comment by jakebailey** [comment]
https://github.com/microsoft/TypeScript/pull/59170 has been merged, so this should be fixed in the next nightly (and I plan to get it backported to 5.5).

**Comment by gcanti** [comment]
Solved by upgrading to latest typescript (v5.5.4) https://github.com/Effect-TS/effect/pull/3360

@jakebailey thank you!
